drivers/sensor: lis2dh: Extended functionality

RubenBoth opened the pull request:
Added support for Wake-up Threshold, and power down functionality as well as some sample functions.

RubenBoth committed:
LIS3DH sample application based on the LIS2DH sample

Signed-off-by: Duma0 <rubenboth@live.nl>

RubenBoth committed:
CONFIG_LIS2DH_ODR_RUNTIME works

Signed-off-by: Duma0 <rubenboth@live.nl>

RubenBoth committed:
Added hardcoded high-pass filter mode selection options, should be turned into KConfig options later.

Signed-off-by: Duma0 <rubenboth@live.nl>

RubenBoth committed:
Implemented Wake-Up threshold and duration as kconfig options. Also added some sample functions in the sample application for testing purposes.

Signed-off-by: Duma0 <rubenboth@live.nl>

RubenBoth committed:
Added power down functionality which is activated by setting ODR to 0, also added High-pass filter kconfig options but that is stil W.I.P.

Signed-off-by: Duma0 <rubenboth@live.nl>

RubenBoth committed:
Added input validation for the Wake Treshold kconfig option

Signed-off-by: Duma0 <rubenboth@live.nl>

RubenBoth committed:
Rebased to integrate code changes from main branch.

Signed-off-by: Duma0 <rubenboth@live.nl>

abelNoelleC commented on drivers/sensor/lis2dh/Kconfig:
```diff
@@ -138,4 +138,41 @@ config LIS2DH_ODR_9_LOW
+	bool "Normal mode"
+
+config LIS2DH_HPM_REFERENCE
+	bool "Difference from reference register"
```
what does this mean?

RubenBoth replied:
This is one of the high pass filter modes. I didn't get around to filling it out more as the high pass filter dropped in priority. If i don't get around to it i will remove it.

abelNoelleC replied:
For the time being I would advise to either remove it or to clearly indicate that this is WIP with a comment or something like that

abelNoelleC commented on drivers/sensor/lis2dh/lis2dh.c:
```diff
@@ -147,10 +147,49 @@ static int lis2dh_acc_odr_set(const struct device *dev, uint16_t freq)
 	if (odr < 0) {
 		return odr;
+	}else if (odr == 0 && data->powered_down == false){
+		/* check if high resolution mode is enabled as it is not allowed to
```
is this the best place to set the low power mode? (if required for the steps in this function, please explain why, otherwise please see if a pm-function can be used)

RubenBoth replied:
The primary way to select the operation mode (low-power / normal / high res) is done through the kconfig. This PR's added functionality is a way for the application to make the sensor go into a full power down mode. 
When the driver sets the ODR to 0 (which the datasheet describes as 'Power down mode') it tries to save as much power as possible by also putting the sensor in low power mode (regardless of what is selected in the kconfig) and disabling all axis as well. 
When the ODR is then set to a value larger than 0 the driver restores the previous (kconfig selected) operating mode and re-enables the axis. 

This functionality could be put in a dedicated function but this integration felt more streamlined to me. Besides that the datasheet's description of 'power down mode' is tied with the ODR, so retaining that connection in the code felt fitting.

abelNoelleC replied:
As we discussed in the meeting on Tuesday, I'm still not sure if doing it like this is the best way. The drivers have device power management capabilities, and if you still have time it would be very beneficial to look into this. Also including stuff like turning off axes, which can greatly impact the power consumption (but this could potentially fit better in the init function).
I understand that the ODR and power down mode are interlinked, so if possible please think about the best solution keeping also the above things in mind

RubenBoth replied:
The power down functionality is now moved to a dedicated function. The option to enable/disable individual axis is on my todo list.

abelNoelleC commented on drivers/sensor/lis2dh/lis2dh.c:
```diff
@@ -261,13 +300,105 @@ static const struct sensor_driver_api lis2dh_driver_api = {
+		    cfg->bus_name, 1 << (LIS2DH_FS_IDX + 1),
+		    LIS2DH_ODR_IDX, (uint8_t)LIS2DH_LP_EN_BIT, lis2dh->scale);
+
+	/* enable accel measurements and set power mode and data rate */
```
So are by default all axis used?

RubenBoth replied:
Yes, the driver currently doesn't have functionality to select specific axis, so its all or nothing.

abelNoelleC replied:
Please state this clearly in a comment and add that as a todo for the driver (somewhere where you're creating an overview of the functionalities and limitations of the driver's current state, like at the beginning of the driver or the README etc)

abelNoelleC commented on drivers/sensor/lis2dh/lis2dh.c:
```diff
@@ -261,13 +300,105 @@ static const struct sensor_driver_api lis2dh_driver_api = {
 	.channel_get = lis2dh_channel_get,
 };
 
+int lis2dh_reg_init(const struct device *dev)
```
why did you create a separate function for this? (so it can be called from the application for something? please specify specific reason in a comment)

RubenBoth replied:
I isolated it so it could be re-called after a reset as the reset function only resets the sensors registers and not the device binding etc that is initialized at startup. Do you mean comments in code or in the PR?

abelNoelleC replied:
Comment in code

abelNoelleC commented on samples/sensor/lis3dh_test/README.rst:
```diff
@@ -0,0 +1,75 @@
+Requirements
+************
+
+This sample uses the LIS2DH, ST MEMS system-in-package featuring a 3D
```
specify more specific requirements (only lis2dh, or also other version like lis3dh; certain microcontroller that has specific peripheral pins? etc)

RubenBoth replied:
I haven't touched this file yet, it's currently an exact copy of the README from the existing LIS2DH sample application.

abelNoelleC replied:
Thought so :), then see this as a reminder to not forget it ;)

abelNoelleC commented on samples/sensor/lis3dh_test/README.rst:
```diff
@@ -0,0 +1,75 @@
+   :goals: build flash
+   :compact:
+
+Building on nucleo_l476rg with IKS01A2 shield
```
do you use this board?

abelNoelleC commented on drivers/sensor/lis2dh/lis2dh.h:
```diff
@@ -272,5 +321,4 @@ int lis2dh_acc_slope_config(const struct device *dev,
 
-
-#endif /* __SENSOR_LIS2DH__ */
+#endif /* __SENSOR_LIS2DH__ */
```
check if the blank lines are added to the files where required

RubenBoth replied:
How would i check this? I put the blank line back for now.

abelNoelleC replied:
if you go to the 'Files Changed' tab in the PR, you'll see some red symbols (circle with a stripe) at the end of some files, indicating the lack of a blank line

abelNoelleC commented on samples/sensor/lis3dh_test/prj.conf:
```diff
@@ -0,0 +1,13 @@
+CONFIG_STDOUT_CONSOLE=y
```
check the other merged PRs for an example for specifying which configs are already included in the board config but just added here to be sure, and which ones aren't

Duma0 replied:
Done

RubenBoth replied:
[Commit](https://github.com/abelsensors/zephyr/commit/a583d21a309f806e0aa53a75107c493e0a623df2)
Is this correct?

abelNoelleC replied:
Yes, maybe also make a section for debugging/logging stuff if you're going to separate that from the other two, but this is what I meant

abelNoelleC reviewed (commented):
Added some comments for an initial review, but I'm aware this is WIP (please also add the WIP label in the PR title)

RubenBoth committed:
Separated configs by whether they are already included by board or not.

Signed-off-by: Duma0 <rubenboth@live.nl>

abelkarapatis commented on drivers/sensor/lis2dh/lis2dh.c:
```diff
@@ -124,9 +124,9 @@ static int lis2dh_freq_to_odr_val(uint16_t freq)
-	if (freq == 0U) {
-		return -EINVAL;
-	}
+	//if (freq == 0U) {
```
Please remove comment

abelNoelleC reviewed (commented):
Responded to some comments

RubenBoth committed:
Moved the powering down functionality to a dedicated function.
Moved the activity threshold configuration from kconfig to custom sensor attributes.
Added a lis2dh.h file in the include directory to make the lis2dh's functions accessible.

Signed-off-by: Duma0 <rubenboth@live.nl>

RubenBoth committed:
drivers/sensor: lis2dh: Added runtime axis selection functionality.
Can be selected in kconfig.

Signed-off-by: Duma0 <rubenboth@live.nl>

RubenBoth committed:
drivers/sensor: lis2dh: Added basic support for PM state controll

Signed-off-by: Duma0 <rubenboth@live.nl>

RubenBoth committed:
drivers/sensor: lis2dh: Polished PM state controll implementation

Signed-off-by: Duma0 <rubenboth@live.nl>

RubenBoth committed:
drivers/sensor: lis2dh: Added ability to set operation mode at runtime

Signed-off-by: Duma0 <rubenboth@live.nl>

abelkarapatis commented on drivers/sensor/lis2dh/lis2dh.c:
```diff
@@ -83,7 +83,7 @@ static int lis2dh_sample_fetch(const struct device *dev,
 {
 	struct lis2dh_data *lis2dh = dev->data;
 	size_t i;
-	int status;
```
Is there a reason you changes status to ret?  Does it serve a purpose?

RubenBoth replied:
Some functions use ret, others used status. I changed all status to ret to make the code more consistent. It doesn't serve any functional purpose.

abelkarapatis commented on drivers/sensor/lis2dh/lis2dh.c:
```diff
@@ -114,7 +113,7 @@ static int lis2dh_sample_fetch(const struct device *dev,
 }
 
-#ifdef CONFIG_LIS2DH_ODR_RUNTIME
+#if defined(CONFIG_LIS2DH_ODR_RUNTIME)
```
Again why did you change it ? The reason I am asking is the less changes you have the less comments you will get from the community

RubenBoth replied:
Mostly code consistency.

abelkarapatis commented on drivers/sensor/lis2dh/lis2dh.c:
```diff
@@ -133,25 +134,25 @@ static int lis2dh_freq_to_odr_val(uint16_t freq)
+	int ret;
 	uint8_t value;
-	struct lis2dh_data *data = dev->data;
+	struct lis2dh_data *lis2dh = dev->data;
```
Try to do only relevant changes

abelkarapatis commented on drivers/sensor/lis2dh/lis2dh.c:
```diff
@@ -165,13 +166,215 @@ static int lis2dh_acc_odr_set(const struct device *dev, uint16_t freq)
+{
+	struct lis2dh_data *lis2dh = dev->data;
+	int ret;
+	uint8_t value;
```
Always initialize your variables

abelkarapatis commented on drivers/sensor/lis2dh/lis2dh.c:
```diff
@@ -165,13 +166,215 @@ static int lis2dh_acc_odr_set(const struct device *dev, uint16_t freq)
+			}
+		}
+		ret = lis2dh->hw_tf->update_reg(dev, LIS2DH_REG_CTRL1,
+									(uint8_t)LIS2DH_LP_EN_BIT_MASK,
```
I would take care on your formatting a bit. This extra tabs are a bit too much

RubenBoth replied:
My IDE displays tabs shorter then github. Ill remove some tabs.

abelkarapatis commented on drivers/sensor/lis2dh/lis2dh.c:
```diff
@@ -165,13 +166,215 @@ static int lis2dh_acc_odr_set(const struct device *dev, uint16_t freq)
+	if (lis2dh->pm_state == PM_DEVICE_STATE_ACTIVE) {
+#if defined(CONFIG_LIS2DH_OPER_MODE_HIGH_RES)
+		int ret;
+		uint8_t value;
```
Again some initialization

abelkarapatis commented on drivers/sensor/lis2dh/lis2dh.c:
```diff
@@ -165,13 +166,215 @@ static int lis2dh_acc_odr_set(const struct device *dev, uint16_t freq)
+
+	if (lis2dh->pm_state == PM_DEVICE_STATE_LOW_POWER) {
+		
+		uint8_t axes = 0;
```
Good one :)

abelkarapatis commented on drivers/sensor/lis2dh/lis2dh.c:
```diff
@@ -207,17 +410,69 @@ static int lis2dh_acc_range_set(const struct device *dev, int32_t range)
+	int ret;
+
+	if ((uint16_t)attr == SENSOR_ATTR_LIS2DH_ACT_TH) {
+		uint8_t range_g, reg_val;
```
Always try to initialize

abelkarapatis commented on drivers/sensor/lis2dh/lis2dh.c:
```diff
@@ -207,17 +410,69 @@ static int lis2dh_acc_range_set(const struct device *dev, int32_t range)
+		range_g = 2 * (1 << ((LIS2DH_FS_MASK & reg_val)
+				      >> LIS2DH_FS_SHIFT));
+
+		act_th_ums2 = val->val1 * 1000000 + val->val2;
```
Add a comment why you are doing this

abelkarapatis commented on drivers/sensor/lis2dh/lis2dh.c:
```diff
@@ -207,17 +410,69 @@ static int lis2dh_acc_range_set(const struct device *dev, int32_t range)
 {
-	switch (attr) {
-#ifdef CONFIG_LIS2DH_ACCEL_RANGE_RUNTIME
+	switch ((uint16_t)attr) {
```
Is there a reason you are casting?  The enum is by default an int value

RubenBoth replied:
This has to do with adding custom sensor attributes. If the value is not casted it cannot be compared to two different enumerators

abelkarapatis replied:
Then my advice would be not to use 2 different enumerators as this can lead to a lot of confusion. Try to place every in one enum. Having an enum helps the compiler check which cases there should be

abelkarapatis reviewed (commented):
Nice work

abelNoelleC commented on samples/sensor/lis3dh_test/boards/stm32l1_disco.overlay:
```diff
@@ -0,0 +1,13 @@
+        reg = <0x18>;
+    	irq-gpios= <&gpiob 8 GPIO_ACTIVE_HIGH>, <&gpiob 8 GPIO_ACTIVE_HIGH>;
+    };
+};
```
mind the missing blank lines at the end of these files (check the red circle with stripe)

abelNoelleC commented on samples/sensor/lis3dh_test/prj.conf:
```diff
@@ -0,0 +1,18 @@
+### Debugging ###
+CONFIG_LOG=y
+CONFIG_LOG_MODE_IMMEDIATE=y
+CONFIG_CBPRINTF_FP_SUPPORT=y
```
missing blank line

RubenBoth committed:
drivers/sensor: lis2dh: Replace code with existing function

Signed-off-by: Duma0 <rubenboth@live.nl>

abelNoelleC reviewed (commented):
Please let us know when the PR is ready for a more thorough review, for now it's just two small comments that I forgot to actually submit :)

RubenBoth committed:
driver/sensor: lis2dh: Added readme.md as well as some minor fixes

Signed-off-by: Duma0 <rubenboth@live.nl>
